File: src/clave/runtime/proposal.py
"""The Python half of the inference boundary.

The format is stated once, in `crates/clave-safety/contract/proposal.md`. This
module encodes it and refuses to send a proposal it already knows the runtime
will refuse, because a producer that ships garbage and reads the complaint back
has learned nothing the check here could not have told it.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from clave.errors import ClaveError
from clave.taxonomy import BY_ID

PROPOSAL_VERSION = 1
"""The wire version this build speaks.

A runtime that meets a version it does not implement refuses the whole message
rather than reading fields under the wrong meaning.
"""

STOP = b""
"""The stop signal: a datagram with no version and no fields."""


class ProposalError(ClaveError):
    """A proposal the runtime would refuse, caught before it is sent."""


@dataclass(frozen=True)
class Proposal:
    """One pick inference proposes, before any check has run.

    Attributes:
        object_id: Identity of the object, carried unchanged into the decision.
        material_class: Taxonomy identifier of the form `M-NN`.
        confidence: Classifier confidence, 0.0 to 1.0 inclusive.
        point: Proposed pick point in belt frame meters.
        yaw_radians: Proposed rotation about the belt normal.
        reference_time_nanos: Instant the pose is predicted for.
        window_start_nanos: Earliest instant the object is reachable.
        window_end_nanos: Latest instant the object is reachable.
    """

    object_id: int
    material_class: str
    confidence: float
    point: tuple[float, float, float]
    yaw_radians: float
    reference_time_nanos: int
    window_start_nanos: int
    window_end_nanos: int

    def __post_init__(self) -> None:
        """Refuse a proposal the runtime would refuse.

        Raises:
            ProposalError: Naming the field that is wrong.
        """
        if self.material_class not in BY_ID:
            raise ProposalError(
                f"material class {self.material_class!r} is not in the taxonomy"
            )
        if not 0.0 <= self.confidence <= 1.0:
            raise ProposalError(
                f"confidence {self.confidence} is outside 0.0 to 1.0 inclusive"
            )
        if self.window_end_nanos < self.window_start_nanos:
            raise ProposalError(
                f"window {self.window_start_nanos} to {self.window_end_nanos} "
                "ends before it starts"
            )
        if not (
            self.window_start_nanos
            <= self.reference_time_nanos
            <= self.window_end_nanos
        ):
            raise ProposalError(
                f"reference time {self.reference_time_nanos} falls outside the window"
            )

    def encode(self) -> bytes:
        """Render as one datagram.

        Returns:
            The UTF-8 JSON the runtime reads.
        """
        x, y, z = self.point
        return json.dumps(
            {
                "version": PROPOSAL_VERSION,
                "object_id": self.object_id,
                "material_class": self.material_class,
                "confidence": self.confidence,
                "x_meters": x,
                "y_meters": y,
                "z_meters": z,
                "yaw_radians": self.yaw_radians,
                "reference_time_nanos": self.reference_time_nanos,
                "window_start_nanos": self.window_start_nanos,
                "window_end_nanos": self.window_end_nanos,
            },
            sort_keys=True,
        ).encode("utf-8")


@dataclass(frozen=True)
class Outcome:
    """What the runtime did with one proposal.

    Attributes:
        verdict: One of `accepted`, `overridden` or `refused`.
        object_id: The object, when one was read.
        channel: The channel the policy resolved, when one was.
        reject_reason: Why it was routed to reject, when it was.
        check: The geometric check that failed, when one did.
        reason: What the decoder reported, when the proposal was refused.
    """

    verdict: str
    object_id: int | None = None
    channel: int | None = None
    reject_reason: str | None = None
    check: str | None = None
    reason: str | None = None

    @property
    def accepted(self) -> bool:
        """Whether a decision was published for this proposal."""
        return self.verdict == "accepted"

    @property
    def overridden(self) -> bool:
        """Whether the safety layer refused the proposed point."""
        return self.verdict == "overridden"

    @classmethod
    def decode(cls, payload: bytes) -> Outcome:
        """Read one outcome datagram.

        Args:
            payload: The JSON the runtime sent.

        Returns:
            The outcome.

        Raises:
            ProposalError: If the payload is not an outcome.
        """
        try:
            read: dict[str, Any] = json.loads(payload)
        except json.JSONDecodeError as error:
            raise ProposalError(
                f"the runtime sent something unreadable: {error}"
            ) from (error)
        verdict = read.get("verdict")
        if verdict not in {"accepted", "overridden", "refused"}:
            raise ProposalError(f"the runtime reported an unknown verdict {verdict!r}")
        return cls(
            verdict=verdict,
            object_id=read.get("object_id"),
            channel=read.get("channel"),
            reject_reason=read.get("reject_reason"),
            check=read.get("check"),
            reason=read.get("reason"),
        )

File: crates/clave-safety/src/proposal.rs
//! The inference boundary.
//!
//! A proposal is what Python sends and Rust receives. It is the only place the
//! two languages meet, and nothing machine-learned crosses it.
//!
//! The wire format is JSON, one object per datagram, and it differs on purpose
//! from the CBOR a published decision uses. A proposal is internal, crossed at
//! most tens of times a second, and is the thing an engineer reads while
//! debugging; a decision is the contract another project compiles against.
//! Readability wins on one side and stability on the other.
//!
//! `contract/proposal.md`, beside this crate, states the fields, their units and the version rule,
//! so a consumer implements against that rather than against this source.

use clave_decision::{
    BeltPoint, Confidence, MaterialClass, MonotonicNanos, ObjectId, PickPose, PickWindow,
};
use serde::Deserialize;

use crate::error::SafetyError;

/// The wire version this build implements.
///
/// A proposal carrying any other version is refused rather than interpreted.
/// The version moves whenever a field is added, removed, or changes meaning.
pub const PROPOSAL_VERSION: u64 = 1;

/// Enough of the payload to read the version, ignoring every other field.
///
/// Decoding happens in two passes so that an unrecognized version is refused
/// on the version alone. A single pass would fail on whichever field the new
/// version changed, reporting a type error where the truthful answer is that
/// the producer is speaking a language this build does not.
#[derive(Debug, Deserialize)]
struct Versioned {
    version: u64,
}

/// The payload as it appears on the wire.
#[derive(Debug, Deserialize)]
struct Wire {
    object_id: u64,
    material_class: String,
    confidence: f32,
    x_meters: f64,
    y_meters: f64,
    z_meters: f64,
    yaw_radians: f64,
    reference_time_nanos: i64,
    window_start_nanos: i64,
    window_end_nanos: i64,
}

/// One pick proposed by inference, before any check has run.
///
/// Every field a decision needs is carried, so the safety layer estimates
/// nothing the model already estimated and invents nothing the model did not.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Proposal {
    object: ObjectId,
    class: MaterialClass,
    confidence: Confidence,
    pose: PickPose,
    window: PickWindow,
}

impl Proposal {
    /// Decodes a proposal from one datagram.
    ///
    /// The version is read first, so an unrecognized one is refused before any
    /// other field is interpreted. Every coordinate is checked to be finite
    /// here, which is what lets the geometric checks be plain arithmetic.
    ///
    /// # Errors
    ///
    /// Returns [`SafetyError::UnsupportedVersion`] when the payload names a
    /// version this build does not implement, [`SafetyError::Malformed`] when
    /// a field is missing, of the wrong type, or not finite,
    /// [`SafetyError::UnknownClass`] when the material class is outside the
    /// taxonomy, and [`SafetyError::Contract`] when the window, the confidence
    /// or the pose reference time is one the published contract refuses.
    ///
    /// Every one of those is returned to the caller. A datagram nobody can
    /// read is not a reason to stop a conveyor.
    pub fn decode(bytes: &[u8]) -> Result<Self, SafetyError> {
        let versioned: Versioned =
            serde_json::from_slice(bytes).map_err(|error| SafetyError::Malformed {
                reason: error.to_string(),
            })?;
        if versioned.version != PROPOSAL_VERSION {
            return Err(SafetyError::UnsupportedVersion {
                found: versioned.version,
            });
        }

        let wire: Wire = serde_json::from_slice(bytes).map_err(|error| SafetyError::Malformed {
            reason: error.to_string(),
        })?;
        let class = from_taxonomy_id(&wire.material_class)?;
        for (name, value) in [
            ("x_meters", wire.x_meters),
            ("y_meters", wire.y_meters),
            ("z_meters", wire.z_meters),
            ("yaw_radians", wire.yaw_radians),
        ] {
            if !value.is_finite() {
                return Err(SafetyError::Malformed {
                    reason: format!("{name} is not a finite number"),
                });
            }
        }

        let window = PickWindow::new(
            MonotonicNanos::new(wire.window_start_nanos),
            MonotonicNanos::new(wire.window_end_nanos),
        )?;
        let reference = MonotonicNanos::new(wire.reference_time_nanos);
        if !window.contains(reference) {
            return Err(SafetyError::Malformed {
                reason: format!(
                    "reference_time_nanos {} falls outside the proposed window",
                    wire.reference_time_nanos
                ),
            });
        }

        Ok(Self {
            object: ObjectId::new(wire.object_id),
            class,
            confidence: Confidence::new(wire.confidence)?,
            pose: PickPose::new(
                BeltPoint::new(wire.x_meters, wire.y_meters, wire.z_meters),
                wire.yaw_radians,
                reference,
            ),
            window,
        })
    }

    /// Returns the identity the tracker assigned to the object.
    #[must_use]
    pub const fn object(self) -> ObjectId {
        self.object
    }

    /// Returns the material class inference assigned.
    #[must_use]
    pub const fn class(self) -> MaterialClass {
        self.class
    }

    /// Returns the confidence behind that class.
    #[must_use]
    pub const fn confidence(self) -> Confidence {
        self.confidence
    }

    /// Returns the pose inference proposes to pick at.
    #[must_use]
    pub const fn pose(self) -> PickPose {
        self.pose
    }

    /// Returns the window inference believes the object is reachable in.
    #[must_use]
    pub const fn window(self) -> PickWindow {
        self.window
    }
}

/// Returns the class with this taxonomy identifier.
///
/// The identifier is matched rather than the variant name, because the
/// identifier is what `docs/waste-taxonomy.md` marks append-only and what the
/// Python side of the boundary already carries.
fn from_taxonomy_id(id: &str) -> Result<MaterialClass, SafetyError> {
    MaterialClass::ALL
        .into_iter()
        .find(|class| class.taxonomy_id() == id)
        .ok_or_else(|| SafetyError::UnknownClass { id: id.to_owned() })
}
